README.md
# About

PrettyQL is a Clojure DSL for Factual's API. It's for when you want to use the world's most powerful programming language to query the world's most powerful open data platform. ;-)

Here's a Pretty statement that finds restaurants near a lat lon, where each restaurant delivers dinner, sorted by distance:

```clojure
(select restaurants-us  
  (around {:lat 34.06021 :lon -118.4183 :miles 3})
  (where              
    (= :meal_deliver true)
    (= :meal_dinner true))
  (order :$distance)    
  (limit 3))
```

# Installation

PrettyQL is hosted at [Clojars](http://clojars.org/pretty-ql). Add this to your project dependencies:

````clojure
[pretty-ql "1.0.4"]
````

Now you can use Pretty on the REPL:

````clojure
(use 'pretty.core)
````

Or in your ns declaration:

````clojure
(require [pretty.core :as pretty])
````

# Authentication

Before your Pretty queries can run, you must provide your Factual API key and secret:

````clojure
(init! "YOUR-KEY" "YOUR-SECRET")
````

# select

<tt>select</tt> takes a table name and an optional set of clauses, such as a <tt>where</tt> clause. Evaluating <tt>select</tt> will run the query against Factual and return a sequence of results hash-maps.

The simplest example:

````clojure
(select places)
````

Another simple example:

````clojure
(select places (where (like :name "starbucks*")))
````

# where

The <tt>where</tt> clause allows you to specify row filters. Here's an example Pretty statement with a tasty <tt>where</tt> clause:

````clojure
; Find restaurants in LA with valid websites and a good rating,
; that won't make me dress too nice:
(select restaurants-us
  (where
    (= :locality "los angeles")
    (not-in :attire ["formal" "smart casual" "business casual"])
    (not-blank :website)
    (>= :rating 2.5)))
````

## supported filter logic

* in
* not-in
* like
* not-like
* search
* blank
* not-blank
* =
* not=
* >
* <
* >=
* <=

# fields

You can specify the fields you want returned using <tt>fields</tt>, like this:

````clojure
(select places
  (fields :name :locality :website))
````

# order

You can order your results, that is sort rows, by using <tt>order</tt>. For example:

````clojure
(select places
  (order :name))
````

Example of ordering by name ascending, locality descending, using explicit :asc and :desc modifiers:

````clojure
(select places
  (order :name:asc :locality:desc))
````

# search (Full Text Search)

Example:

````clojure
(select places
  (search "starbucks"))
````

# limit and offset

You can limit the returned results with <tt>limit</tt>, like:

````clojure
(select restaurants-us (limit 12))
````

You can page through results using <tt>order</tt>, <tt>limit</tt> and <tt>offset</tt>, like:

````clojure
(select places
  (order :name:asc :locality:desc)
  (offset 20)
  (limit 10))
````

# around (Geo Proximity Filter)

You can use <tt>around</tt> to limit your results to be within a geographic radius of a lat lon coordinate. For example:

````clojure
; Find places near Factual:
(select places
  (around {:lat 34.06021 :lon -118.4183 :miles 3}))
````

# Schema

You can get the schema for a specific table like this:

````clojure
(schema restaurants-us)
````

# Query Composition

Pretty provides support for composing queries. You can define a query without running it, using <tt>select*</tt>. Later you can create new queries based on that query, and run them at anytime.

For example, imagine you want to define a base query that finds U.S. restaurants that have valid owners and telephones:

````clojure
(def base (-> (select* "restaurants-us")
              (where
                (not-blank :owner)
                (not-blank :tel))))
````

Running that query is as easy as using <tt>exec</tt>:

````clojure
(exec base)
````

But you can also define a new query that builds off of <tt>base</tt>. For example, let's define a new query that uses <tt>base</tt> but adds a filter for non-null websites, and also adds a sort based on website:

````clojure
(def websites (-> base
                (where 
                  (not-blank :website))
                  (order :website)))
````

You can run the <tt>websites</tt> query anytime with <tt>exec</tt>:

````clojure
(exec websites)
````

You can go on to create a new query that is is similar to the <tt>websites</tt> query but, say, adds a limit clause:

````clojure
(def a-few-sites (-> websites (limit 3)))
````

And of course, you can run that at anytime with <tt>exec</tt>.

And the previous <tt>base</tt> and <tt>websites</tt> queries are unchanged, so you can continuing building new queries off of them as well.


# Resolve

PrettyQL provides simple and direct support for Factual's Resolve feature. The function is called <tt>resolve</tt>, and it expects values as a hashmap, where keys are valid attributes for the Resolve schema, and values are the values on which to match.

For example:

````clojure
(resolve {:name "ino", :latitude 40.73, :longitude -74.01})
````

# License

The use and distribution terms for this software are covered by the
Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
which can be found in the file LICENSE.html at the root of this distribution.
By using this software in any fashion, you are agreeing to be bound by
the terms of this license.
You must not remove this notice, or any other, from this software.


CHANGELOG.md
## 1.0.4

 * Added init! and deprecated pretty!

src/pretty/core.clj
(ns pretty.core
  (:require [funnyplaces.api :as fun]
            [clojure.walk :as walk]
            [clojure.string :as str]))

;; --- pretty helpers ---

;; assumes s always like "term*"
;; PS: Laura says: iloveyou
(defn like-str->m [s]
  {:$bw (apply str (butlast s))})

;; assumes s always like "term*"
(defn not-like-str->m [s]
  {:$nbw (apply str (butlast s))})

(defn pred
  "Creates the most common predicate spec. %1 is field name, %2 is val.
   Example use in fn lookup table:
   '> (pred :$gt)
   Example final result when used to interpret (... (> :rank 8.5) ...):
   {:rank {:$gt 8.5}}"
  [pred-name]
  `#(hash-map %1 {~pred-name %2}))

;; --- pretty fn translation tables ---

(def keywords
  {'where 'pretty.core/where
   'fields 'pretty.core/fields
   'order 'pretty.core/order
   'search 'pretty.core/search
   'offset 'pretty.core/offset
   'limit 'pretty.core/limit
   'circle 'pretty.core/circle
   'around 'pretty.core/around})

(def preds {'and 'pretty.core/+and  
            'or 'pretty.core/+or
            'search 'pretty.core/search
            'like 'pretty.core/+like
            'not-like 'pretty.core/+not-like
            'blank 'pretty.core/+blank
            'not-blank 'pretty.core/+not-blank
            '= (pred :$eq)
            'not= (pred :$neq)
            '> (pred :$gt)
            '>= (pred :$gte)
            '< (pred :$lt)
            '<= (pred :$lte)
            'in (pred :$in)
            'not-in (pred :$nin)})

;; --- pretty fns ---

(defn fields [q & forms]
  ;;;(update-in q [:select] into (map name forms))
  (assoc q :select (str/join "," (map name forms)))
  )

(defn order [q & forms]
  (assoc q :sort (str/join "," (map name forms))))

(defn +or [& clauses]
  {:$or (vec clauses)})

(defn +and [& clauses]
  {:$and (vec clauses)})

(defn search
  "Supports 2 variants of full text search:

   1) At top level of query, so FTS across row, like:
   (select ... (search \"myterm\"))
   arg1 will be the query map, arg2 the search term.

   2) Within the where clause, so for a specific field, like:
   (select ... (where ... (search :tel \"myterm\")))
   arg1 will be the field name, arg2 the searh term."
  [arg1 arg2]
  (if (map? arg1)
    (assoc arg1 :q arg2)
    {arg1 {:$search arg2}}))

(defn +like [field term]
  {field (like-str->m term)})

(defn +not-like [field term]
  {field (not-like-str->m term)})

(defn +blank [field]
  {field {:$blank true}})

(defn +not-blank [field]
  {field {:$blank false}})

(defn limit [q limit]
  (assoc q :limit limit))

(defn offset [q offset]
  (assoc q :offset offset))

(defn circle
  "Adds a geo proximity filter to query q.
   Expects a hash-map describing a circle, like:
    {:center [40.73 -74.01]
     :meters 5000}

   Supports one of :miles or :meters"
  [q circ]
  (let [meters (or (:meters circ) (* (:miles circ) 1609.344))
        center {:$center (:center circ) :$meters meters}]
    (assoc q :geo {:$circle center})))

(defn around
  "Adds a geo proximity filter to query q.
   Expects a hash-map describing a circle, like:
   {:lat 34.06021 :lon -118.4183 :miles 3}
  
   Only supports :miles"
  [q {:keys [lat lon miles]}]
  (circle q {:center [lat lon] :miles miles}))

;; --- pretty DSL ---

(defn init!
  "Establishes your oauth credentials for Factual. You only need to
   this once for the lifetime of your application."
  [key secret]
  (fun/factual! key secret))

(defn pretty!
  "DEPRECATED: Use 'init!' instead"
  [key secret]
  {:added "0.0.1"
   :deprecated "1.0.4"}
  (fun/factual! key secret))

(defn exec [query]
  (let [table (:table (meta query))]
    (fun/fetch-q table query)))

(defn select* [table]
  (with-meta
    {}
    {:type :select
     :table table}))

(defn where*
  "Add a where clause to the query. Clause can be either a map or a string, and
  will be AND'ed to the other clauses."
  [query clause]
  (update-in query [:filters] conj clause))

(defmacro where [query & clauses]
  (let [xform (walk/postwalk-replace preds clauses)]
    `(update-in ~query [:filters] merge ~@xform)))

(defmacro select [table & clauses]
  (let [xform (walk/postwalk-replace keywords clauses)]
    `(let [query# (-> (select* ~(name table)) ~@xform)]
       (exec query#))))

(defmacro schema [table]
  `(fun/schema ~(name table)))

(defn resolve
  "Direct support for Resolve. values must be a hashmap, where keys
   are valid attributes for the schema, and values are the values on
   which to match."
  [values]
  (fun/resolve values))